Features/ManageabilityPkg/Universal/IpmiSmbiosTransferDxe/IpmiSmbiosTransferDxe.c

```diff
@@ -0,0 +1,293 @@
+  //
+  Smbios30TableModified->EntryPointStructureChecksum = 0;
+  Smbios30TableModified->EntryPointStructureChecksum =
+    CalculateCheckSum8 ((UINT8 *)Smbios30TableModified, Smbios30TableModified->EntryPointLength);
+
+  SendData = AllocateZeroPool (sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT) + Smbios30Table->TableMaximumSize);
+  CopyMem (SendData, Smbios30TableModified, sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT));
+  CopyMem (SendData + sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT), (UINT8 *)Smbios30Table->TableAddress, Smbios30Table->TableMaximumSize);
+  SendDataSize      = sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT) + Smbios30Table->TableMaximumSize;
+  RemainingDataSize = SendDataSize;
+
+  if (PcdGetBool (PcdSendSmbiosOnChanged)) {
+    SmbiosTransferRequired = DetectSmbiosChange (SendData, SendDataSize);
+
+    if (!SmbiosTransferRequired) {
+      DEBUG ((DEBUG_INFO, "%a: Smbios tables are not changed, skipping transfer to BMC\n", __func__));
+      return;
+    }
+  }
+
+  DEBUG_CODE_BEGIN ();
+  DEBUG ((SMBIOS_TRANSFER_DEBUG, "%a: SMBIOS BINARY DATA OUTPUT\n", __func__));
+  DEBUG ((SMBIOS_TRANSFER_DEBUG, "%a: Table Address: %x\n", __func__, Smbios30Table->TableAddress));
+  DEBUG ((SMBIOS_TRANSFER_DEBUG, "%a: Table Length: %x\n", __func__, Smbios30Table->TableMaximumSize));
+  for (Index = 0; Index < SendDataSize; Index++) {
+    if ((Index % BLOB_MAX_DATA_PER_PACKET) == 0) {
+      DEBUG ((SMBIOS_TRANSFER_DEBUG, "\nIndex:%x ", Index));
+    }
+
+    DEBUG ((SMBIOS_TRANSFER_DEBUG, "%02x ", *(SendData + Index)));
+  }
+
+  DEBUG_CODE_END ();
+
+  Status = IpmiBlobTransfer->BlobOpen ((CHAR8 *)PcdGetPtr (PcdBmcSmbiosBlobTransferId), BLOB_TRANSFER_STAT_OPEN_W, &SessionId);
+  if (EFI_ERROR (Status)) {
+    if (Status == EFI_UNSUPPORTED) {
+      return;
+    }
+
+    DEBUG ((DEBUG_ERROR, "%a: Unable to open Blob with Id %a: %r\n", __func__, PcdGetPtr (PcdBmcSmbiosBlobTransferId), Status));
+    goto ErrorExit;
+  }
+
+  for (Index = 0; Index < (SendDataSize / BLOB_MAX_DATA_PER_PACKET); Index++) {
+    Status = IpmiBlobTransfer->BlobWrite (SessionId, Index * BLOB_MAX_DATA_PER_PACKET, SendData, BLOB_MAX_DATA_PER_PACKET);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "%a: Failure writing to blob: %r\n", __func__, Status));
+      goto ErrorExit;
+    }
+
+    SendData           = (UINT8 *)SendData + BLOB_MAX_DATA_PER_PACKET;
+    RemainingDataSize -= BLOB_MAX_DATA_PER_PACKET;
+  }
+
+  ASSERT (RemainingDataSize < BLOB_MAX_DATA_PER_PACKET);
+  if (RemainingDataSize) {
+    Status = IpmiBlobTransfer->BlobWrite (SessionId, Index * BLOB_MAX_DATA_PER_PACKET, SendData, RemainingDataSize);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "%a: Failure writing final block to blob: %r\n", __func__, Status));
+      goto ErrorExit;
+    }
+  }
+
+  Status = IpmiBlobTransfer->BlobCommit (SessionId, 0, NULL);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failure sending commit to blob: %r\n", __func__, Status));
+    goto ErrorExit;
+  }

```

Comment:
The blob commit without polling the blob state does not work for me. 

According to https://github.com/openbmc/phosphor-ipmi-blobs?tab=readme-ov-file#bmcblobcommit-5, 

> The commit operation may exceed the IPMI timeout duration of ~5 seconds (implementation dependant). Callers are expected to poll on BmcBlobSessionStat or BmcBlobStat (as appropriate) until committing has finished. To address race conditions, blobs should not allow concurrent sessions that modify state.

We should continue polling the stat until the blob has been committed.

```

  UINTN   RetryIndex;
  UINT16  BlobState;

  Status = IpmiBlobTransfer->BlobCommit (SessionId, 0, NULL);
  if (EFI_ERROR (Status)) {
    // Poll the stat for the blob committed
    for (RetryIndex = 0, BlobState = 0; RetryIndex < 10; RetryIndex++, BlobState = 0) {
      Status = IpmiBlobTransfer->BlobSessionStat (SessionId, &BlobState, NULL, NULL, NULL);
      if (!EFI_ERROR (Status)) {
        if ((BlobState & BLOB_TRANSFER_STAT_COMMITTED) != 0) {
          DEBUG ((DEBUG_INFO, "%a: Blob committed %r\n", __func__));
          break;
        }

        if ((BlobState & BLOB_TRANSFER_STAT_COMMIT_ERROR) != 0) {
          DEBUG ((DEBUG_ERROR, "%a: Failure sending commit to blob: %r\n", __func__, Status));
          break;
        }
      }

      MicroSecondDelay (200 * 1000); // 200ms
    }
  }

```

Reply:
Thanks for the suggestion. I add this retry in latest commit.